x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/predicate/operator/arithmetic/Arithmetics.java

```diff
@@ -153,25 +157,35 @@ static Number negate(Number n) {
         return Integer.valueOf(Math.negateExact(n.intValue()));
     }
 
-    private static Temporal periodArithmetics(Temporal l, Period r, BiFunction<Temporal, Period, Temporal> f) {
+    private static Temporal periodArithmetics(Temporal l, Period r, IntervalOperation operation) {
         if (l == null || r == null) {
             return null;
         }
 
         if (l instanceof OffsetTime) {
-            r = r.withDays(0).withMonths(0).withYears(0);
+            return l;
+        }
+
+        if (operation == IntervalOperation.ADD) {
+            return l.plus(r);
+        } else {
+            return l.minus(r);
         }
-        return f.apply(l, r);
     }
 
-    private static Temporal durationArithmetics(Temporal l, Duration r, BiFunction<Temporal, Duration, Temporal> f) {
+    private static Temporal durationArithmetics(Temporal l, Duration r, IntervalOperation operation) {
         if (l == null || r == null) {
             return null;
         }
 
         if (l instanceof OffsetTime) {
```

Comment:
You could fold the two methods by passing the duration/period as `TemporalAmount` and then do a type check internally:
```java
if (l instanceof OffsetTime) {

   if (r instanceof Period) {
     return l;
   } else if (r instanceof Duration) {
    r = ((Duration) r).ofMillis(..);
  } else {
     throw new SqlIllegalArgumentException("Unsupported Temporal type [{}]", r);
  }
```

it's not too ugly and avoid the null check and if (operation == ) repetition.

Reply:
Imho since we know if it's a Period or Duration it's "ugly" to merge them into TemporalAmount and then do another `instanceof` (we do it already for the `OffsetTime`).